recipes/mpfr/all/conanfile.py

```diff
@@ -29,148 +34,140 @@ class MpfrConan(ConanFile):
+
+            env = tc.environment()
+            if self.settings.compiler == "msvc":
+                env.define("AR", "lib")
+                env.define("CC", "cl -nologo")
+                env.define("CXX", "cl -nologo")
+                env.define("LD", "link")
+                env.define("NM", "dumpbin -symbols")
+                env.define("OBJDUMP", ":")
+                env.define("RANLIB", ":")
+                env.define("STRIP", ":")
+            tc.generate(env) # Create conanbuild.conf
+            tc = AutotoolsDeps(self)
+            tc.generate()
 
     def _extract_makefile_variable(self, makefile, variable):
-        makefile_contents = tools.load(makefile)
-        match = re.search("{}[ \t]*=[ \t]*((?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\\\"))*(?:\\\\\n(?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\"))*)*)\n".format(variable), makefile_contents)
+        makefile_contents = load(self, makefile)
+        match = re.search(f'{variable}[ \t]*=[ \t]*((?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\\\"))*(?:\\\\\n(?:(?:[a-zA-Z0-9 \t.=/_-])|(?:\\\"))*)*)\n', makefile_contents)
         if not match:
-            raise ConanException("Cannot extract variable {} from {}".format(variable, makefile_contents))
+            raise ConanException(f"Cannot extract variable {variable} from {makefile_contents}")
         lines = [line.strip(" \t\\") for line in match.group(1).split()]
         return [item for line in lines for item in shlex.split(line) if item]
 
     def _extract_mpfr_autotools_variables(self):
-        makefile_am = os.path.join(self._source_subfolder, "src", "Makefile.am")
+        makefile_am = os.path.join(self.source_folder, "src", "Makefile.am")
         makefile = os.path.join("src", "Makefile")
         sources = self._extract_makefile_variable(makefile_am, "libmpfr_la_SOURCES")
         headers = self._extract_makefile_variable(makefile_am, "include_HEADERS")
         defs = self._extract_makefile_variable(makefile, "DEFS")
         return sources, headers, defs
 
-    @contextlib.contextmanager
-    def _build_context(self):
-        if self.settings.compiler == "Visual Studio":
-            with tools.vcvars(self):
-                env = {
-                    "AR": "lib",
-                    "CC": "cl -nologo",
-                    "CXX": "cl -nologo",
-                    "LD": "link",
-                    "NM": "dumpbin -symbols",
-                    "OBJDUMP": ":",
-                    "RANLIB": ":",
-                    "STRIP": ":",
-                }
-                with tools.environment_append(env):
-                    yield
-        else:
-            yield
-
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+        with chdir(self, self.source_folder):
+            command = "./autogen.sh"
+            if os.path.exists(command):
+                self.run(command)
+        apply_conandata_patches(self)
```

Comment:
I think it's better to patch first and then regenerate the configure scripts so eventual changes to `configure.ac`
 are applied to `configure`.
```suggestion
        apply_conandata_patches(self)
        with chdir(self, self.source_folder):
            command = "./autogen.sh"
            if os.path.exists(command):
                self.run(command)
        apply_conandata_patches(self)
```

Reply:
Re-ordered as suggested.